### DISCUSSION TITLE
v2.0.0 code

### DISCUSSION THREAD

**Discussion by 877dev** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#41] [General]
Hey there, 

Finally got some time to check out this new version v2.0.0. Looks quite different, must've taken a lot of work! I've only had a quick scroll through but seems good.

Couple of thoughts:

1) Did not compile for me, I had to rename "BWC_globals.h" to "BWC_8266_globals.h" in the library. It could be changed elsewhere I'm sure, just a spelling error.

2) My pins are different to yours due to a different layout, I edited "BWC_8266.h", no biggie. But would be nice to be able to change them by just setting one flag somewhere.

3) MQTT credentials, can't see where these are inputted? There's no option on the web interface like before, and I don't see reference to "myMqttIP"or "myMqttPort" anywhere else. So of course I'm getting -2 connection error. Are you not using mqtt anymore?

4) Is there any keep alive function? I think in read that someone unlocked the device every day in on of the ticker functions?

**Comment by visualapproach** [comment]
Hi! Thank you, yes it took some time :-)

1. Doh! Corrected now. I changed my local filenames while I was trying to port the project to ESP32*. I made some minor improvements that I wanted to upload and I forgot that file. 
2. Yes, the pinout is kind of hidden now. I may add parameters to the class.
3. It's in the example files "globals.h". Put BWC_v2.0.0 folder in "libraries" and open the examples from Arduino IDE menu. Then you just edit the globals.h tab.
4. You can easily add a repeating command to the queue from the config webpage (or via mqtt). For instance "set pump ON now and repeat every day"

I suggest you go with the example "basic with mqtt and gpio". If you are going to use the GPIOs you have to uncomment those parts in the code.

Cheers

**Reply by hlaci1984** [reply]
Hello. I'm right here. please help me, I can't find the codes for arduino esp8299 Wi-FI anywhere. Lay-Z-spa model S100101

**Reply by hlaci1984** [reply]
e-mail: horvathladislav84@gmail.com

**Reply by visualapproach** [reply]
> Hello. I'm right here. please help me, I can't find the codes for arduino esp8299 Wi-FI anywhere. Lay-Z-spa model S100101

Hello. [start right here](https://github.com/visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA/blob/master/bwc-manual.pdf)

**Reply by hlaci1984** [reply]
Thanks. where can the ESP8266 setup be downloaded?

**Comment by 877dev** [comment]
Ah yes I see the "globals.h" tab, somehow missed it! :)

I'll have a little play after this week when work is quieter.
Nice job! 👍🏻

**Comment by visualapproach** [comment]
\* I tried to see if I could port to ESP32 to learn something. It didn't go well ;-(
Cores panicked and it kept crashing until I gave up.

**Comment by visualapproach** [comment]
2. Done

**Comment by 877dev** [comment]
Nice thank you. It looks like they can be defined in one line, which is excellent. 

I've been looking at your code, specifically "BWC_8266.cpp". Am I correct that if no parameters are passed to bwc.begin, via the main sketch, it is classed as (void) and the default values are used? And if parameters ARE passed, then it skips the void section and goes to the second function?

If so, I've never really understood that void could be used that way..


UPDATE: It's working so I guess this is how it works :)

**Comment by visualapproach** [comment]
> Ok so now transferring over my node red flow to the new format, I have it working reading the values from mqtt.
> 
> But I cannot send mqtt commands currently, and I notice that even commands from the web interface do not work for me. So maybe that is work in progress?
> 
> I tried values of true and 1, and XTIME of 0 and current unix time...
> 
> e.g.
> {"CMD":2,"VALUE":1,"XTIME":1606161550,"INTERVAL":0}
> or
> {"CMD":2,"VALUE":true,"XTIME":1606161550,"INTERVAL":0}
> or
> {"CMD":2,"VALUE":1,"XTIME":0,"INTERVAL":0}

Hi, It should be working as is. Do you send to default topic "**BW_2.0.0/command**"? With payload as above, which is correctly formatted. (I might have changed the topic base name.)

> I've been looking at your code, specifically "BWC_8266.cpp". Am I correct that if no parameters are passed to bwc.begin, via 
> the main sketch, it is classed as (void) and the default values are used? And if parameters ARE passed, then it skips the void 
> section and goes to the second function?
> 
> If so, I've never really understood that void could be used that way..

You are right about the parameters - The "void" is actually not the thing here. It just tells the compiler that no parameters are expected. It can actually be omitted in many cases.
`void myfunction(void){} `
is the same as 
`void myfunction(){} `
When writing a class you can make several functions with the same name, but with different parameter types or count, and the compiler will choose the one version that matches the parameters given. It's called overloading a function.
So in the function 
`void bwc.begin(void)`
I set the pins to default value. But in 
`void bwc.begin(int, int, int, int, int, int, int)` I set pins according to the parameters.

**Comment by 877dev** [comment]
> Hi, It should be working as is. Do you send to default topic "**BW_2.0.0/command**"? With payload as above, which is correctly formatted. (I might have changed the topic base name.)

Turns out I had a spelling mistake :-)  I deleted my comment but you replied too soon lol. It's working now.

The Command Queue on the web interface never shows anything for me though, which threw me off.

> It's called overloading a function.

Overloading a function, that is interesting thanks for the info. And thanks for adding that in, makes my life easier with my dodgy pins.

I'll work on the node red flows next, for hooking my Blynk app to control the pump. Before it was toggled by sending i.e. 0x5 for power, now it needs a different payload per state. Pretty easy to hook up with nodes, but I will try and write a function to make it more eloquent. 

There's no up and down control anymore, but I can replace that with a slider. And the unlock button is no longer needed it seems. Finally timer button, but who uses the timer button anyway! :)

**Comment by visualapproach** [comment]
Haha, I'm glad it worked out. But I wonder why the web interface didn't. Did you upload the new data folder to LittleFS?
And yes, I thought the UP/DOWN, ON/OFF and LOCK was obsolete now. And the factory timer was too confusing to use anyway.

I set up my node red such that when the set temperature is reached I get an email :-). Or when temp is below 10C. I must be old, thinking such things are cool...

**Comment by 877dev** [comment]
Yes I did upload LittleFS and have the new web interface. Not a big deal really. 

Any reason why power button is missing from the commands? For example if someone turns it off outside, does the code handle that and power it up?

**Comment by visualapproach** [comment]
Yes I removed it because you don't really care if it's on or not. The off state just seems to dim the display and annoys people when they wanna do something like lower the temp or so. Same with the lock. So if you send a command, let's say "set temp to 40", the esp first turns on the power and then unlocks the device (if necessary) and then simulates UP until display reads 40.

**Comment by visualapproach** [comment]
Btw, the web interface queue on the 'config' page refreshes every 4 seconds. Maybe you was a little impatient?
Or is it the main page that doesn't work for you?

**Comment by 877dev** [comment]
No it definitely does not update, this is the page I am referring to:

![image](https://user-images.githubusercontent.com/46575282/100273818-4f0e6080-2f55-11eb-94bf-2246c6eca695.png)

I click "Add command to queue" and it never appears..

**Comment by visualapproach** [comment]
Okay, a couple of thoughts;
Is that a normal electric price in your area?
It may be clear to you already, but just in case - Adding a command with xtime in the past, or "now" will execute immidiately and will be deleted from the queue before it pops up on the page. If you add an interval more than zero, it will first execute, then add the interval time to the command. Meaning it should show up in the queue. Likewise for single events in the future.
You could check the browser console (Chrome: shift+ctrl+I) what is sent/received.

**Comment by 877dev** [comment]
> Is that a normal electric price in your area?

I assume so, it's £0.1295 per kwh, the interface only seems to allow two decimal places, so 0.13. Also there is a daily charge of £0.195 on top of that. 
Why is that good or bad? :-)



> Adding a command with xtime in the past, or "now" will execute immidiately

That wasn't obvious at first, but I figured it out. Didn't know that "now" works also, even better.


> If you add an interval more than zero, it will first execute, then add the interval time to the command. Meaning it should show up in the queue. Likewise for single events in the future.

Does not work for me, I'm using the web interface and selecting a few minutes in the future, also sending interval of 8600. This is console output, never looked in there before..

Example of changing temp to 40 degrees:
![image](https://user-images.githubusercontent.com/46575282/100787280-50cf9c80-340b-11eb-8ba5-4e2265b21e2c.png)

**Comment by visualapproach** [comment]
Price- I was just curious. I checked my bill and my rate is 0.58 SEK (around £0.05) but there is a fixed price on top of that too. Difficult to compare.

"Now" meant current timestamp at the time you send the command. Not the literal string "Now".

The errors indicate that you don't have the correct config.HTML file. The line numbers don't match mine.
Download latest files and via Ip/upload.html choose the newest html files.

Wish you luck

**Comment by 877dev** [comment]
> The errors indicate that you don't have the correct config.HTML file. The line numbers don't match mine.
> Download latest files and via Ip/upload.html choose the newest html files.

Strange, I used the latest files from github. I just manually uploaded each file and have the same result. Also rebooted ESP.

Could it be something I have inadvertently changed in the sketch? Are you sure the uploaded data files are the correct ones? :-)

**Comment by visualapproach** [comment]
Errors would occur if you changed the sketch to send data not matching the expected format. However those errors would occur at specific lines in the web page's source. Now, if you, instead of looking at the console tab in your browser,  go to the SOURCE tab. Select config.html. Here you can see what is ACTUALLY on your board. Compare with the config.html in latest example/data folder.

This is from the config.html file on this github. You see the "{LEN:0}" string is printed at line 112, as opposed to line 105 in your picture.
![cfg1](https://user-images.githubusercontent.com/22987250/101151011-58688e80-3621-11eb-8e6b-7adf816bd98e.png)

By clicking source, here we see that the console.log is at the line 112.

![cfg2](https://user-images.githubusercontent.com/22987250/101151315-c0b77000-3621-11eb-8680-9e79990a8f63.png)

**Comment by 877dev** [comment]
Hey, just looking at this now coincidentally. Give me 10 mins to check.. :)

Update: 

In Sources I get this at line 112:

![image](https://user-images.githubusercontent.com/46575282/101152599-38ce6780-361b-11eb-8cad-238f9eab2217.png)

Also before you posted I was looking at 'http://192.168.1.135/eventlog.txt' and I get a 404: file not found error, is that an issue?


Still getting an error when adding to queue:
![image](https://user-images.githubusercontent.com/46575282/101153099-de81d680-361b-11eb-9774-bca0f9e7e226.png)

 and in sources:

![image](https://user-images.githubusercontent.com/46575282/101153374-3f111380-361c-11eb-951a-0c95dd59e06f.png)


cmdq.txt show `{"LEN":0}` when there is no command, and `{"LEN":1}` when I add to command queue (although it does not show in the web interface queue).

**Comment by visualapproach** [comment]
Hmm, it expects a message like:  {"LEN":1,"CMD":[2],"VALUE":[0],"XTIME":[1607685540],"INTERVAL":[0]}
but only gets {"LEN":1}. Hence trying to parse CMD etc will throw a UNDEFINED error.
So check your code if it's sending all info. Original code responsible for this:

```
//this is the sketch file
//response to /getcommands/
void handleGetCommandQueue() { // reply with json document
  if (server.method() != HTTP_POST) {
    server.send(405, "text/plain", "Method Not Allowed");
  } else {
    String jsonmsg = bwc.getJSONCommandQueue();
    server.send(200, "text/plain", jsonmsg);
  }
}
```
 
```
//this is the .cpp file
String BWC::getJSONCommandQueue(){
  // Allocate a temporary JsonDocument
  // Don't forget to change the capacity to match your requirements.
  // Use arduinojson.org/assistant to compute the capacity.
  //feed the dog
  ESP.wdtFeed();
  DynamicJsonDocument doc(1024);
  // Set the values in the document
  doc["LEN"] = _qCommandLen;
  for(int i = 0; i < _qCommandLen; i++){
	  doc["CMD"][i] = _commandQ[i][0];
	  doc["VALUE"][i] = _commandQ[i][1];
	  doc["XTIME"][i] = _commandQ[i][2];
	  doc["INTERVAL"][i] = _commandQ[i][3];
  }

  // Serialize JSON to file
  String jsonmsg;
  if (serializeJson(doc, jsonmsg) == 0) {
    jsonmsg = "{\"error\": \"Failed to serialize message\"}";
  }
  return jsonmsg;
}
```

**Comment by 877dev** [comment]
Very strange, my sketch and .cpp are exactly the same...

**Comment by visualapproach** [comment]
Yes, strange it is. In the console window, enter
edited:
{"CMD":2,"VALUE":1,"XTIME":0,"INTERVAL":0}
and press enter. See if that works.

regarding eventlog.txt - No it is not a problem. That file only exists if you create it by using saveEventlog() but that is not recommended (by me) because it was a debug function which will eventually fill up your flash.

**Comment by visualapproach** [comment]
@877dev you should look at IP/cmdq.txt
Maybe something weird is saved there.

**Comment by 877dev** [comment]
> Yes, strange it is. In the console window, enter
> edited:
> {"CMD":2,"VALUE":1,"XTIME":0,"INTERVAL":0}
> and press enter. See if that works.

Seems to be accepted ok, but no bubble turn on, or show up in the command queue.
![image](https://user-images.githubusercontent.com/46575282/101247494-e6bc3d80-3711-11eb-980f-b47b9680edca.png)

From the webinterface, if I send a command in the past or current time, it works. 
**But for any future event is shows the error...**

cmdq.txt just show `{"LEN":0}`
Whenever I add a command to the queue, it shows `{"LEN:1}` and gives the error.

**Comment by visualapproach** [comment]
OK, do you use my code or custom? I know you changed the pinouts, but anything else?

**Comment by 877dev** [comment]
I use the gpio code, changed pinouts and mqtt reporting timer to 5 seconds. But I tried with default 600s too. Not anything else I can think of. Also re-uploaded LittleFS just in case. 

Just tried it with fresh install:
Copy latest github files over library files
Open example "basic" sketch from ArduinoIDE
Add my custom pins 
Flash sketch
Flash LittleFS

Same error message :(

**Comment by visualapproach** [comment]
Same code and same Files. Shouldn't be it.
What differs?
- Other hardware
- Dependancy libraries. Try updating ArduinoJSON, and other libraries.
- Time and timezone
- Browser? I use chrome/safari both works
- Board settings in arduino IDE?

**Comment by visualapproach** [comment]
Does it work when pressing the pump display buttons? Or via MQTT?

**Comment by 877dev** [comment]
Sorry for delay, I've worked all weekend :(

OK so I checked all of this, same error I'm afraid.

Other hardware **NONE**
Dependancy libraries. Try updating ArduinoJSON, and other libraries. **UPDATED ALL**
Time and timezone **TRIED 0 AND 1**
Browser? I use chrome/safari both works **TRIED CHROME AND IE**
Board settings in arduino IDE? **SEEM CORRECT**
Does it work when pressing the pump display buttons? **YES ALL BUTTON WORK**
 Or via MQTT? **"NOW" COMMANDS WORK, BUT FUTURE COMMANDS DO NOT**

Also tried basic sketch, only thing I changed is the pins, same error. 



It's a shame no-one else can weigh in, as it might be just something on my end..
But not to worry, I can control the pump via mqtt and use timer nodes etc..

**Comment by 877dev** [comment]
One thing I noticed is the time is "American" format, i.e. MM/DD/YYYY

![image](https://user-images.githubusercontent.com/46575282/101346641-c19a0d00-3880-11eb-9e2e-fd04acdb2734.png)

Whereas in the config it is EU format DD/MM/YYYY

![image](https://user-images.githubusercontent.com/46575282/101346773-f8702300-3880-11eb-9aac-e9b59f5191c2.png)

Might explain why future commands are not working for me..
But I think the main page sometimes changes to EU format after a while, or am I imagining that?

**Comment by visualapproach** [comment]
No worries, I've been busy trying to repair my 2:nd broken pump :(. Stupid impeller needs to be exactly at the right position, not even half a mm off axis...
Afaik (not home now) the esp is sending timestamps and the browser is converting to where it thinks you are.
We will check another thing. Call you back.

**Comment by 877dev** [comment]
If it's too much trouble don't worry, it's more an annoyance it's not working than anything:)

Have you got / did you have Grafana working? I notice that when set to target of 40, the heater does not come back on until the actual temp drops down to 38. I wonder if that is by design or in your code?

I might setup a node red flow to kick the heater back in when it drops to 39, only during certain hours. Getting a bit chilly over here 😁￼

![image](https://user-images.githubusercontent.com/46575282/101377080-4e58c100-38a9-11eb-8652-cd8edbb889b8.jpeg)

PS the left most one is where I manually switched it back on, the middle one was the tub being used. I'm such a data geek lately!

**Comment by visualapproach** [comment]
Check in your library folder. Remove any old folder containing this project. Any #include statement may use the first matching file it finds, wich may be in another folder.
Yes my Grafana is working. 
I didn't code it to switch at 38/40 so it must be by design. I think I read that the material in the kit was made for 38 C, so I think they made it so that it will not exceed 40 degrees.

**Comment by visualapproach** [comment]
Also, have you tried it on another 8266? On the bench, not connected to something other than USB.

**Comment by 877dev** [comment]
I only have one library folder. I will try it on another esp8266, I have a spare pump head unit I can plug it into.
Little things like this are quite annoying for some reason, messes with my ocd  ;-)

**Comment by visualapproach** [comment]
I know the feeling

**Comment by 877dev** [comment]
Spare esp8266 not connected to anything, flashed fresh from basic sketch. Same issue:

![image](https://user-images.githubusercontent.com/46575282/101602083-72291d80-39f5-11eb-84d0-2cb50986459f.png)

**Comment by 877dev** [comment]
Serial monitor output while error occurs, is the "Disconnected" anything?

![image](https://user-images.githubusercontent.com/46575282/101602572-3cd0ff80-39f6-11eb-9d38-086ca1c9fd04.png)

**Comment by visualapproach** [comment]
I'll have a look after work. LED is probably on One of the data pins.

**Comment by visualapproach** [comment]
Ok, so I had a look. Downloaded repository from github and recompiled basic example. Uploaded LittleFS files and it worked like a charm. There is something fishy though. Why does your browser say line 110 and mine 112?! It might be nothing. Maybe it's because it is getting an error.

![config_serial](https://user-images.githubusercontent.com/22987250/101668831-57849200-3a51-11eb-8a22-9faf4180708d.png)

![config_console](https://user-images.githubusercontent.com/22987250/101668851-5fdccd00-3a51-11eb-9efb-3aad8a677c53.png)

![config_source](https://user-images.githubusercontent.com/22987250/101668861-62d7bd80-3a51-11eb-9654-b27e5d4110d7.png)

**Comment by visualapproach** [comment]
Also line 146/147 is different. It's like you have a different html file, but the received msg shouldn't be just {"LEN":1}. it makes no sense.
Edit BWC_8266.cpp like this:

```
String BWC::getJSONCommandQueue(){
  // Allocate a temporary JsonDocument
  // Don't forget to change the capacity to match your requirements.
  // Use arduinojson.org/assistant to compute the capacity.
  //feed the dog
  ESP.wdtFeed();
  DynamicJsonDocument doc(1024);
  // Set the values in the document
  doc["LEN"] = _qCommandLen;
  for(int i = 0; i < _qCommandLen; i++){
	  doc["CMD"][i] = _commandQ[i][0];
	  doc["VALUE"][i] = _commandQ[i][1];
	  doc["XTIME"][i] = _commandQ[i][2];
	  doc["INTERVAL"][i] = _commandQ[i][3];
  }

  // Serialize JSON to file
  String jsonmsg;
  if (serializeJson(doc, jsonmsg) == 0) {
    jsonmsg = "{\"error\": \"Failed to serialize message\"}";
  }
Serial.print("This is debugging: ");
Serial.println(jsonmsg);  //<<---------------------- insert this
  return jsonmsg;
}
```

**Comment by 877dev** [comment]
![image](https://user-images.githubusercontent.com/46575282/101678302-ab917580-3a55-11eb-9b6c-add370567d34.png)

**Reply by visualapproach** [reply]
Okay, now edit these functions in the same file.

```
String BWC::getJSONCommandQueue(){
  // Allocate a temporary JsonDocument
  // Don't forget to change the capacity to match your requirements.
  // Use arduinojson.org/assistant to compute the capacity.
  //feed the dog
  ESP.wdtFeed();
  DynamicJsonDocument doc(1024);
  // Set the values in the document
  doc["LEN"] = _qCommandLen;
  for(int i = 0; i < _qCommandLen; i++){
	  doc["CMD"][i] = _commandQ[i][0];
	  doc["VALUE"][i] = _commandQ[i][1];
	  doc["XTIME"][i] = _commandQ[i][2];
	  doc["INTERVAL"][i] = _commandQ[i][3];
          Serial.print("CMD: "); Serial.print(_commandQ[i][0]);
          Serial.print(", VALUE: "); Serial.print(_commandQ[i][1]);
          Serial.print(", XTIME: "); Serial.print(_commandQ[i][2]);
          Serial.print(", INTERVAL: "); Serial.println(_commandQ[i][3]);
  }

  // Serialize JSON to file
  String jsonmsg;
  if (serializeJson(doc, jsonmsg) == 0) {
    jsonmsg = "{\"error\": \"Failed to serialize message\"}";
  }

  return jsonmsg;
}
```

```
bool BWC::qCommand(uint32_t cmd, uint32_t val, uint32_t xtime, uint32_t interval) {
	//handle special commands
	if(cmd == RESETQ){
		_qButtonLen = 0;
		_qCommandLen = 0;
		_saveCommandQueue();
		return true;
	} 

	//add parameters to _commandQ[rows][parameter columns] and sort the array on xtime.
	int row = _qCommandLen;
	if (_qCommandLen == MAXCOMMANDS) return false;
	//sort array on xtime
	for (int i = 0; i < _qCommandLen; i++) {
		if (xtime < _commandQ[i][2]){
			//insert row at [i]
			row = i;
			break;
		}
	}	
	//make room for new row
	for (int i = _qCommandLen; i > (row); i--){
		_commandQ[i][0] = _commandQ[i-1][0];
		_commandQ[i][1] = _commandQ[i-1][1];
		_commandQ[i][2] = _commandQ[i-1][2];
		_commandQ[i][3] = _commandQ[i-1][3];
	}
	//add new command
	_commandQ[row][0] = cmd;
	_commandQ[row][1] = val;
	_commandQ[row][2] = xtime;
	_commandQ[row][3] = interval;
	_qCommandLen++;
       Serial.print("qCommand: ");
      Serial.print(cmd); Serial.print(" "); Serial.print(val); Serial.print(" "); Serial.print(xtime); Serial.print(" "); Serial.println(interval);
	delay(0);
	_saveCommandQueue();
	return true;
}
```

**Comment by 877dev** [comment]
Hmm does not want to compile with those changes ..

https://pastebin.com/gsrXVdwi

**Reply by visualapproach** [reply]
I typed that code directly in the chat so I may have introduced a typo or something, but someone was having a similar issue awhile ago. IIRC a reinstall of arduino ide helped.

**Comment by 877dev** [comment]
I’ll try that now on another machine. I think a list of needed libraries would be helpful, maybe I missed it. 
Here's what I found I needed on a fresh install:

**Basic sketch:**
ArduinoJSON (Benoit Blanchon)
ESPDateTime https://github.com/mcxiaoke/ESPDateTime
WebSockets https://github.com/Links2004/arduinoWebSockets
WiFiManager https://github.com/tzapu/WiFiManager


**MQTT sketch (in addition to basic):**
PubSubClient https://github.com/knolleary/pubsubclient


**LittleFS install:**
https://arduino-esp8266.readthedocs.io/en/latest/filesystem.html#uploading-files-to-file-system

**Comment by 877dev** [comment]
Success!! A fresh install on my laptop and it's working as expected. Sorry if I have created a headache for you, it seems it is a problem with the Arduino installation on my main PC.....

![image](https://user-images.githubusercontent.com/46575282/101768154-205bc280-3add-11eb-901f-82fc29f293b2.png)

**Comment by visualapproach** [comment]
Lovely! No worries, it’s good to know what is causing it. Even better that it’s not my bad programming ;)

**Comment by hlaci1984** [comment]
Hello. can you send me the esp8266 wifi setup. thanks

visualapproach ***@***.***> ezt írta (időpont: 2023. jún.
10., Szo, 18:57):

**Reply by visualapproach** [reply]
It's all here for you and everybody else to download. If you by "Wi-Fi setup" mean the code for this project.

**Comment by hlaci1984** [comment]
do you happen to have a video? step by step


LADISLAV HORVATH ***@***.***> ezt írta (időpont: 2023.
jún. 11., V, 20:58):

> yes I meant the project code. I just can't find where and what to download?
>>
>

**Reply by visualapproach** [reply]
https://googlethatforyou.com?q=how%20to%20download%20github%20project

**Comment by hlaci1984** [comment]
there is how to download it. I just don't understand what exactly? there
are many files there and it is not clear to me what the ESP8266 needs to
copy.

**Comment by hlaci1984** [comment]
I think it has everything. I just can't configure the ESP8266

**Reply by visualapproach** [reply]
Did you read the manual?

**Comment by hlaci1984** [comment]
Yes. I downloaded it. why is what I bought not good? something doesn't fit?

**Comment by hlaci1984** [comment]
I don't really understand it. I think he already realized that. any advice
would be appreciated.

**Reply by visualapproach** [reply]
I put a lot of effort into writing the manual and I'm not sure I can express myself any clearer. Your requests are very generic. You haven't said anything about what pump you have, what you have or haven't done, or what specific problem you have or which page in the manual you don't understand. When I asked if you read it you said you downloaded it. You need to read it also. I can't read minds and I don't have time to figure out what you mean.
You are welcome to post any specific questions in an issue related to your specific problem, or if there are none, open a new issue or discussion.
